theano/tensor/nnet/nnet.py

```diff
@@ -414,39 +473,37 @@ class Softmax(gof.Op):
     :math:`\\varphi(\\mathbf{x})_j =
     \\frac{e^{\mathbf{x}_j}}{\sum_{k=1}^K e^{\mathbf{x}_k}}`
     where :math:`K` is the total number of neurons in the layer. This
-    activation function gets applied row-wise.
-
+    activation function gets applied on the selected axis (Default axis is the last).
+    This function get applied when axis is a scalar and uses a fast C optimization.
     """
 
     nin = 1
     nout = 1
     __props__ = ()
+    params_type = ParamsType(axis=Scalar('int8'))
+
+    def __init__(self, axis=-1):
+        self.axis = axis
 
     def make_node(self, x):
         x = tensor.as_tensor_variable(x)
-        if x.type.ndim not in (1, 2) \
-                or x.type.dtype not in tensor.float_dtypes:
-            raise ValueError('x must be 1-d or 2-d tensor of floats. Got %s' %
-                             x.type)
-        if x.ndim == 1:
-            x = tensor.shape_padleft(x, n_ones=1)
-
         return Apply(self, [x], [x.type()])
 
-    def perform(self, node, input_storage, output_storage):
+    def perform(self, node, input_storage, output_storage, param):
         x, = input_storage
-        e_x = np.exp(x - x.max(axis=1)[:, None])
-        sm = e_x / e_x.sum(axis=1)[:, None]
+        axes = param.axis
+        # Apply softmax on the specified dimension
+        e_x = np.exp(x - x.max(axis=axes, keepdims=True))
+        sm = e_x / e_x.sum(axis=axes, keepdims=True)
         output_storage[0][0] = sm
 
     def L_op(self, inp, outputs, grads):
         x, = inp
         g_sm, = grads
-        return [softmax_grad(g_sm, outputs[0])]
+        return [SoftmaxGrad(self.axis)(g_sm, outputs[0])]
```

Comment:
This way, every time that perform is called a new ```SoftmaxGrad``` object is created. Doesn't it increase the memory usage and time?

Reply:
We realized when timing similar Ops that caching them did not really make a difference.

Reply:
@bordesf could you also add tests for that?